Amazon1.Step5-자동검색완성

HTMLhead opened the pull request:
* 피드백반영 사항

1. 값을 할당하며 Object.assign(this, {원하는 객체}) 를 사용하여 this내부에 원하는 객체값들이 복사되도록 만들었습니다.
2. qs의 default parameter를document로 수정하였습니다.
3. header up.scss파일에 em, rem을 도입해보았습니다.

* 변경사항

1. carousel에 사용하던 xhr을 fetch로 수정하였습니다.

* 코드맥락

1. 이번에 만드는 자동검색완성 또한 app.js 에 search_autocorrect에 여러가지 값을 받아서 이모저모 설정할 수 있도록 만들었습니다.
2. setTimeout으로 debounce를 만들었습니다. keydown이벤트가 감지되면 이전에 있던 setTImeout을 제거해주고 keyup이벤트가 감지되면 다시 setTimeout을 설정하도록 만들었습니다.
3. enter, arrowup, arrowdown이 입력되면 바로 실행될 수 있도록 제작하였습니다.
4. 함수마다 꼭 한가지의 일을 하도록 만드려고 노력했습니다.
5. li의 element에 data-{something}을 넣어서 li마다 값을 가지고 있도록 만들고, 클릭 되었을 때, url을 만들어 주소창에 입력되어 form 데이터에 submit이 되도록 만들었습니다.
6. form데이터에 input이 있으면 enter를 눌렀을 때, 자동으로 submit이 되는 문제가 있어서 html파일에 `onsubmit='return false'`를 추가했습니다. 
7. form 데이터 내부에 select, input, button 태그들이 존재하도록 만들었습니다.

짜는동안 코드를 일반적인 방법으로 짠 것 같지가 않았습니다..보통 이렇게 짜는지 궁금합니다. 리뷰 부탁드립니다! 🆘

HTMLhead committed:
reset.scss 수정

reset.scss를 scss문법에 맞게 수정

HTMLhead committed:
util.js 의 qs변경 및 변경사항 적용

기본값을 document로 가질 수 있도록 함수 수정

HTMLhead committed:
app.js 에 option쪽 추가

autoEventStopContainer를 추가해서 마우스가 올라갔을때 이벤트가 멈추는 것 추가

HTMLhead committed:
carousel-middle.js 파일 object.assign사용해서 변경

object.assign을 이용해서 모든객체를 this에 병합해서 복사하고, 그 복사한 값들을 이용해서
내부를 변경해 줄 수 있도록 수정

HTMLhead committed:
carousel-middle.js 수정완료

HTMLhead committed:
scrollEvent-Sticky.js 파일 object.assign을 이용해 수정

init()함수 내부에 makeIO함수를 넣어서 초기화해주는 느낌이 들도록 수정, constructor함수
내부를 최대한 간단히 하도록 만듬.
또한, makeIO함수에 stickyLayer와 hiddenLayer부분을 인자로 받아서 사용하도록 수정

HTMLhead committed:
headerup.scss head-search-container쪽 icon수정

마진값 rem으로 수정

HTMLhead committed:
headerup.scss 의 head search-selector수정

값 rem으로 모두 변경

HTMLhead committed:
headerUP.scss 완성

rem으로 모두 수정

HTMLhead committed:
headerDown.scss 수정, head-nav-location부분

depth에 맞게 수정하고, px부분 모두 rem으로 수정

HTMLhead committed:
headerDown.scss 수정 head-nav부분

rem으로 모두 병경하고, depth sass식으로 변경

HTMLhead committed:
headerDown.scss rem과 depth로 수정완료

HTMLhead committed:
intersectionObserver이름변경

본래에 있는 js파일들과 표기법을 같게 변경하여 적용

HTMLhead committed:
search-autocorrect.js파일 추가

HTMLhead committed:
index.html에 태그추가

검색어 자동완성을 위해서 필요한 태그를 만들어서 html코드에 추가

HTMLhead committed:
autocorrect의 디자인 완성

HTMLhead committed:
util.js 파일, carousel-middle.js에 fetch적용

fetch를 적용함으로써 필요없어진 getAjax함수를 없애고, fetch로 변경하여 적용

HTMLhead committed:
getAjax함수 제거

HTMLhead committed:
search-autocorrect.js 스켈레톤 코드 완성

HTMLhead committed:
app.js에 상황에 맞게 적용

HTMLhead committed:
getSearchData함수 하드코딩완성

HTMLhead committed:
app.js에 option추가

검색어 창이 띄워졌을 때, 주어진 부분을 투명해지게 해주는 부분 추가, 투명해지는 시간
정하는 옵션 추가

HTMLhead committed:
search-autocorrect.js와 wholeBody.css변경

입력받았을때 주변이 투명해지도록 기능추가

HTMLhead committed:
search-autocorrect.js 에 highlightData 기능 추가, 값 함수로 나눔

하이라이트 되는 색깔추가, 그 단어를 제외한 값 은 하이라이트가 안되도록 제작

HTMLhead committed:
search-autocorrect.js에 makeUrlData, index, autocorrect.scss수정

그냥 문자열을 url문자열로 변경해주는 makeUrlData함수 구현, select, input, button 모두에 form데이터가 필요할것 같아서 form데이터 내부에 넣도록 수정 한 후에, 그에맞게 sass도 수정

HTMLhead committed:
검색어마다 url을 가지도록 제작

HTMLhead committed:
app, search-autocorrect.js 리팩토링

코드를 위에서부터 쉽게 읽기 위해 코드정리, cloakingTransisitonTime이라고 정확하게 명시

HTMLhead committed:
isUpDownArrowOrEnter 함수 완성

up키나 down, enter키가 눌리면 정해진 함수로 들어가며 리턴 시켜주도록 작성

HTMLhead committed:
아래위를 누를때 백그라운드 하이라이트 되도록 수정

HTMLhead committed:
search-autocorrect.js의 화살표엔터키 반응 버그수정

화살표를 밑으로 누를때 인덱스가 끝없이 추가되어 일어나던 버그와 이전의 값이 존
재하지 않아서 위로 올라가고도 계속 인덱스가 마이너스 되던 부분 수정
엔터를 눌렀을 때, 값이 value로 올라가고 검색창이 사라지도록 구현

HTMLhead committed:
enter를 눌렀을 때, submitFormData가 실행되는것 수정

enter를 눌렀는데 마우스 이벤트가 실행되는 문제가 있어서 수정. e.x === 0을 주면 해결이 되긴
하는데 이게 맞는 수정법인지 모르겠음

HTMLhead committed:
search-autocorrect.js 리팩토링 getSearchData

여러가지의 일을 하기에 한가지 일을 함수로 따로 뺌, isInputBlank함수로 input이 비어있는지
체크해주는 기능

HTMLhead committed:
getSearchData에서 resetAutocorrectLists의 기능 따로 분리

HTMLhead committed:
search-autocorrect.js 함수들의 위치 변경

HTMLhead committed:
settimeout을 이용해서 debounce제작

crongro commented on amazonJS/app.js:
```diff
@@ -1,24 +1,38 @@
+  });
+
+  const search_autocorrect = new Search_autocorrect({
+    searchWindow: '.head-search-blank',
```
잘했어요.
이런식으로 selector 문자열을 넣어줘도 됩니다.
아니면 클래스안의 생성자에서 해도 되긴하고요. 

지금처럼 모든 엘리먼트를 다 넣어주기보다는 기준엘리먼트 정도만 넣어주고, 
나머진 그 기준엘리먼트를 기준으로 찾기도 하죠. 
그런데 이렇게 되면 constructor 안에 css selector 명이 하드코딩된것인데, 이게 좋은것인가? 의문이 들긴해요.
재사용 클래스에서는 class이름을 때로는 강제화하기도해요. 이 클래스(컴포넌트)를 쓰려면 오른쪽 버튼의 css class이름은 이거여야해! 라는 식으로요.
(잘 이해 안가면 월요일에 설명할게요)

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -1,28 +1,29 @@
-    this.transitionTime = optionObj.transitionTime;
-    this.transitionPart = optionObj.transitionPart;
-    this.carouselAutoMovingMS = optionObj.carouselAutoMovingMS;
+    Object.assign(this, { elObj, urlObj, optionObj });
```
👍

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -1,28 +1,29 @@
   init() {
-    getAjax(this.handler.bind(this), this.ajaxDataUrl);
+    let ajaxDataUrl = this.urlObj.ajaxDataUrl;
+    fetch(ajaxDataUrl).then(res => {
```
자 같으면 init에서 해야할일은 한줄씩 의미있는 것들을 표현할 거 같아요.
그래서 fetch를 분리할 거 같네요. 이렇게요.

getData() {
  fetch()...
}

init() {
getData();
setTimeout();
setTimeout();
}

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -83,8 +87,9 @@ class Carousel_middle {
-    this.container.addEventListener("mouseout", this.mouseOut.bind(this));
+    this.autoEventStopContainer = qs(this.optionObj.autoEventStopContainer);
+    this.autoEventStopContainer.addEventListener("mouseover", this.mouseOver.bind(this));
+    this.autoEventStopContainer.addEventListener("mouseout", this.mouseOut.bind(this));
```
참고로, 
arrow function을 사용하면 bind가 필요없죠.

crongro commented on amazonJS/scrollEvent-Sticky.js:
```diff
@@ -2,42 +2,43 @@ import { qs } from "./util.js";
+    
+    this.makeIO(qs(this.elObj.header), hiddenLayer, stickyLayer);
 
     closeBtn.addEventListener("click", this.hideLayer.bind(this));
```
이 부분을 init함수 안에서 할것인가? 
아니면 이 3줄을 regEvents() 라는 함수에 넣어서 여기서 그 함수를 부를것인가?  

비슷한 아이들이 보이면 묶어주고 분리하는 것도 좋아요.

crongro commented on amazonJS/scrollEvent-Sticky.js:
```diff
@@ -2,42 +2,43 @@ import { qs } from "./util.js";
 
-  makeIO(root) {
+  makeIO(root, hiddenLayer, stickyLayer) {
     const io = new IntersectionObserver(entries => {
```
오 IntersectionObserver 를 사용!! 잘 이해하고 계시면 더 좋겠습니다.

HTMLhead replied:
엄청 일부분만 자세히 이해하려고 노력했었던 것 같아요. 다른 재미난 것을 만들 때 사용해보면서 더 연습 해 볼 계획입니다.

crongro commented on amazonJS/scrollEvent-Sticky.js:
```diff
@@ -2,42 +2,43 @@ import { qs } from "./util.js";
 
   showHiddenLayer() {
-    this.hiddenLayer.classList.add("prime-member-container-active");
+    qs(this.elObj.hiddenLayer).classList.add("prime-member-container-active");
```
소스코드(함수)에서 css class명과 같은 부분을 잘 분리했는데요. 
여기서는 css class가 하드코딩되어 있어서 약간 아쉽네요.
prime-member-container-active 이부분도 뺄 수  있으면 빼도 됩니다.

이런 이유는 css class이름이 바뀌어도, 중요한 내부 함수를 수정하지 않고, 
좀더 바깥쪽 코드를 수정하게 된다는 점이죠.

crongro commented on amazonJS/search-autocorrect.js:
```diff
@@ -0,0 +1,177 @@
+import { qs } from "./util.js";
```
es modules의 다양한 import/export 방식을 이해하고 계시길.

crongro commented on amazonJS/search-autocorrect.js:
```diff
@@ -0,0 +1,177 @@
+    if (this.isUpDownArrowOrEnter(e.key)) return this.operateCertainKeyEvent(e);
+    this.resetAutocorrectLists.call(this);
+    const inputValue = this.searchWindow.value;
+    if (this.isInputBlank.call(this, inputValue)) return;
```
isInputBlank 에서 사용하는 this 가 달라지나 보군요.
혹시 불필요한것은 아닌지 한번 점검해보세요.

crongro commented on amazonJS/search-autocorrect.js:
```diff
@@ -0,0 +1,177 @@
+  addList(jsonData) {
+    let nowData = "";
+    if (jsonData.suggestions === undefined) return;
+    jsonData.suggestions.forEach(suggestion => {
```
reduce...사용가능한지 확인하기.

crongro commented on amazonJS/search-autocorrect.js:
```diff
@@ -0,0 +1,177 @@
+    let nowData = "";
+    if (jsonData.suggestions === undefined) return;
+    jsonData.suggestions.forEach(suggestion => {
+      nowData += `<li
```
이런 template 코드가  지금보다 좀더  길~어지면 클래스 밖에서 만들고  import 해서 사용하는 것도 고려해볼만하긴해요.
저번에 server작업할때 pug파일을 분리했던 이유랑 같은 셈이죠.

HTMLhead replied:
안그래도 고민을 하긴 했었어요..저번에 chrome Extension을 만들고 나서 누나에게 리뷰를 받았었는데, 그런 문자열 템플릿들을 import해서 사용한다고도 하더라구요. 근데 일단 문자열이 그렇게 길지 않고 정리도 어느정도 되기도 해서 사용했습니다. 나중에 한번 그렇게도 만들어봐야겠네요.

crongro commented on amazonJS/search-autocorrect.js:
```diff
@@ -0,0 +1,177 @@
+  addUrl(e) {
+    const parentTargetData = e.target.parentNode.dataset;
+    const targetData = e.target.dataset;
+    if (e.target.tagName === "SPAN") {
```
e.target을 자꾸 쓰면 변수로 담아서 사용하던가.(const t = e.target)
아니면 e.target만 받으면 destructuring 을 사용해서 파라미터로 쓰던가하시죠.

e.target처럼 반복적인 객체참조 코드는 미미하긴 하지만 느린코드에요. 변수로 저장해서 캐싱효과로 인한 성능향상을 !

crongro commented on amazonJS/search-autocorrect.js:
```diff
@@ -0,0 +1,177 @@
+  }
+  makeUrlKeywords(data) {
+    const urlData = data
+      .split("")
```
오 메서드 체이닝 잘 했네요~ 함수형이죠 이런프로그래밍 방식도요.

crongro commented on amazonJS/search-autocorrect.js:
```diff
@@ -0,0 +1,177 @@
+      list[this.autocorrectListIndex + 1].style = "background-color:#fff;";
+      return;
+    }
+    list[this.autocorrectListIndex].style = "background-color:#f8f8f8;";
```
style 객체에 값을 할당하는 것인데? 되는군요.
저도 이렇게 안해봤는데 신기하네요.
element.style.background  = ""; 
이렇게만 했었는데. 

가급적 classList를 사용할 수 있으면 그렇게 해서, js안에서 css style코드가 직접 표현되지 않도록 하세요. 예를들어서 color 값이 바뀌면.. js를 수정해야 하는 것은 좀 좋지 않은 결과에요.(js코드해서 이상한 오류가 생길 수도 있으니.. 빌드도 다시해야 하는 위험부담도 있고요)

crongro commented on amazonJS/app.js:
```diff
@@ -1,24 +1,38 @@
+import { Search_autocorrect } from "./search-autocorrect.js";
 
 document.addEventListener("DOMContentLoaded", () => {
   const scrollEvent_sticky = new ScrollEvent_sticky({
```
app.js 전체 구성이 좋습니다. 이렇게 많이 짜요.

crongro commented on amazonJS/search-autocorrect.js:
```diff
@@ -0,0 +1,177 @@
+import { qs } from "./util.js";
+
+class Search_autocorrect {
```
하나의 js파일에 하나의 class를 만든것은 좋아요. 
클래스 이름은 (_) 보다는 카멜표기법이 좀더 일반적일수도 있어요. 하지만 이것은 팀또는개인의 룰이죠.

HTMLhead replied:
파일의 이름과 클래스명과의 싱크로를 최대한 맞추고 싶어서 이렇게 만들었어요. 근데 보니까 어색하기도 하네요. 수정하겠습니다.

crongro reviewed (changes requested):
아주 잘 구현했음(진심)

리뷰내용은 꼭 반영하지 않아도 될 것들이 많습니다.

하나의 모듈(Class)안에서의 함수의 역할, 구조, 함수간의 연결부분등 모두 좋아요.

잘했습니다. 

재사용 되는 클래스로 승격시키기(?)에 대해서는 조만간 수업을 할꺼라 그때 필요한 부분을 더 이해하고 깨닫게 되실 예정.

리뷰내용 선택적으로 좀더 수정하면 좋겠습니다.
그리고 모든 코드에 대해서 스스로 설명할 수 있으면 좋겠어요 (이 코드가 가진 의미가 뭐고, 어떻게 동작하고, 왜 이런식으로 구현했는지<어떤 장점을 예상했기에>) 

---

그리고 pr질문에 코멘트를 달자면, 

기본 이벤트를 취소하려면, 
onsubmit='return false' 이렇게 html안에 js코드를 넣기보다는, 
이런 방법으로 js에서 이벤트로 처리해주는 게 좀더 나아요.
이렇게 하면 여기코드에 submit할때 입력값 validation 처리도 여기에 같이 할 수도 있고요.

```javascript
form.addEventListener(“submit”, (e)=> {
  e.preventDefault(); });
```

일반적으로 짜는 것인가? 에 대한 궁금증은 너무 갖지 않고 자신있게 구현하시면 됩니다.
쉽게 이야기하면 돌아가면 일단 잘 한 거에요!

아 그리고 커밋메시지 앞부분에 설명을 요약해서 표현해보세요(fix, change. 등 현재형으로)
참고 :https://medium.com/@nawarpianist/git-commit-best-practices-dab8d722de99

HTMLhead committed:
Feat: 자동완성 기능의 JS파일에 이벤트를 막는 기능 추가 및 적용

원래 html 파일에 submit(return false) 코드를 넣었었는데 html 파일에 JS 코드가 있는 게 자연스럽지 않다
고 판단되어 JS파일로 이동

HTMLhead committed:
Fix: 캐러 샐 기능의 JS 파일에 fetch를 실행해주는 함수를 추가 이후 적용

fetch가 init 함수 내에서 실행되는 것이 어색하다고 느껴져서 함수로 빼서 따로 구현

HTMLhead committed:
Fix: 캐러 샐 기능의 JS파일에 오른쪽 왼쪽 버튼 이벤트 등록을 arrowFunction으로 수정

HTMLhead committed:
Fix: 캐러 샐 기능의 자동으로 움직이는 기능의 setTimeout함수에 arrowFunction사용

HTMLhead committed:
Fix: 캐러 샐 기능의 움직일지 말지 결정하는 기능의 이벤트등록에 화살표 함수 사용

HTMLhead committed:
Chore: 분기점이 되는 JS 파일에 클래스 명을 강제하는 주석 추가

HTMLhead committed:
Refactor: 붙어있는 바 기능의 JS 파일의 init 함수 내부 이벤트 등록을 화살표 함수로 수정

HTMLhead committed:
Style: 붙어있는 바 기능의 JS 파일의 공백제거 및 세미콜론 추가

5번라인 세미콜론 추가, 8번라인 공백제거, 15번라인 공백제거

HTMLhead committed:
Refactor: 붙어있는바 기능의 JS 파일에 이벤트를 등록하는 기능의 함수 추가

HTMLhead committed:
Refactor: 분기점이 되는 JS 파일에 클래스명들을 모두 문자열로 받도록 수정, 적용

HTMLhead committed:
Chore: 분기점이 되는 JS 파일에 주석 제거

HTMLhead committed:
Refactor: 붙어있는 바 기능의 JS파일에 이벤트가 등록되는 요소를 함수에 추가

HTMLhead committed:
Fix: 붙어있는 바 기능의 JS파일에 요소등록 안되던 버그 수정

HTMLhead committed:
Refactor: 붙어있는 바 기능의 JS 파일에 옵서버 API를 사용한 함수 내부에 요소등록

HTMLhead committed:
Refactor: 자동완성기능 JS파일에 쓸모없던 빈값인지 확인하는 함수 제거

HTMLhead committed:
Refactor: 자동완성기능 JS 파일에 리스트 추가함수에 for Each를 reduce로 수정

HTMLhead committed:
Refactor: 자동완성기능 JS 파일에 주소 추가함수에 target을 변수에 담도록 수정

HTMLhead committed:
Refactor: 자동완성기능 JS 파일에 이벤트 등록을 화살표함수로 수정

하면서 debounce가 중복으로 선언되는 문제를 함수를 debouncing으로 이름을 바꿔서 해결

HTMLhead committed:
Feat: 자동완성기능 SCSS 파일에 선택되었을 때 음영 짓도록 하는 클래스 추가

HTMLhead committed:
Refactor: 분기가 되는 JS 파일에 자동완성쪽 option추가 및 적용

HTMLhead committed:
Chore: 자동완성기능 JS 파일에 값을 input에 넣어주는 기능 쪽 주석제거

HTMLhead committed:
Fix: 분기가 되는 JS 파일의 자동완성쪽 option의 clock오타수정

cloak => cloaking

HTMLhead committed:
Chore: 스네이크 표기법을 모두 카멜표기법으로 수정

HTMLhead committed:
Chore: 분기점이 되는 JS 파일의 자동완성기능부분의 객체의 속성이름 변경

HTMLhead commented:
리뷰 해주신 부분 수정해보았습니다. 커밋 메세지에 신경을 많이 써 보았습니다.

다시한번 리뷰 부탁드립니다! 😄

crongro commented on amazonJS/app.js:
```diff
@@ -1,24 +1,46 @@
+  },{
+    ajaxDataUrl: "./jsonData/data.json"
+  },{
+    carouselSize: "230px",
```
230px 이런것도 CSS정보이기 때문에 변경을 염두해둬야해요.
이값을 하드코딩하지 않고, 로딩이후에  계산에 의해서 이 값을 javascript로 추출해 낼 수 있다면 좋겠죠.

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -1,41 +1,46 @@
     this.linkUrlArr = parsedObj.linkArr;
-    this.right.addEventListener("click", this.moveRight.bind(this));
-    this.left.addEventListener("click", this.moveLeft.bind(this));
+    this.right.addEventListener("click", (e) => {this.moveRight()});
```
(물론, bind를 쓰는게 잘못되거나 다른 문제가 있는 건 아닙니다)

crongro reviewed (approved):
잘 수정했음요.

destructuring 을 잘 사용해보면 더 좋겠네요.
